chunkset_tpl.h

```diff
@@ -2,69 +2,64 @@
  * For conditions of distribution and use, see copyright notice in zlib.h
  */
 
+/* Copies a partial chunk, all bytes minus one */
+static inline uint8_t* CSUFFIX(chunkcopy_partial)(uint8_t *out, uint8_t const *from, unsigned len) {
+    int32_t use_chunk16 = sizeof(chunk_t) > 16 && (len & 16);
+    if (use_chunk16) {
+        memcpy(out, from, 16);
+        out += 16;
+        from += 16;
+    }
+    if (len & 8) {
+        memcpy(out, from, 8);
+        out += 8;
+        from += 8;
+    }
+    if (len & 4) {
+        memcpy(out, from, 4);
+        out += 4;
+        from += 4;
+    }
+    if (len & 2) {
+        memcpy(out, from, 2);
+        out += 2;
+        from += 2;
+    }
+    if (len & 1) {
+        *out++ = *from++;
+    }
+    return out;
+}
+
 /* Returns the chunk size */
-Z_INTERNAL uint32_t CHUNKSIZE(void) {
+Z_INTERNAL uint32_t CSUFFIX(chunksize)(void) {
     return sizeof(chunk_t);
 }
 
-/* Behave like memcpy, but assume that it's OK to overwrite at least
-   chunk_t bytes of output even if the length is shorter than this,
-   that the length is non-zero, and that `from` lags `out` by at least
-   sizeof chunk_t bytes (or that they don't overlap at all or simply that
-   the distance is less than the length of the copy).
-
-   Aside from better memory bus utilisation, this means that short copies
-   (chunk_t bytes or fewer) will fall straight through the loop
-   without iteration, which will hopefully make the branch prediction more
-   reliable. */
-Z_INTERNAL uint8_t* CHUNKCOPY(uint8_t *out, uint8_t const *from, unsigned len) {
+/* Copy memory in chunks when possible and avoid writing beyond legal output */
+static inline uint8_t* CSUFFIX(chunkcopy_static)(uint8_t *out, uint8_t const *from, unsigned len, uint8_t *safe) {
```

Comment:
I have thought about changing the last argument from `safe` to `unsigned left`. Any preference @mtl1979?

Reply:
I personally think `left` would simplify the code and reduce risk of off-by-one errors.